#12421 ProccessMonitor: make startProcess() catch exceptions raised by spawn…

juanpeha opened the pull request:
…Process().

## Scope and purpose

Fixes #12421

spawnProcess() might fail, ProcessMonitor should be able to deal with such failures and attempt to restart the failed process.

Example:

```
km@stable ~/pyng/bin $ ./run.py                                                                    
While calling system event trigger handler
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/twisted/internet/base.py", line 518, in _continueFiring
    callable(*args, **kwargs)
  File "/home/km/pyng/lib/pyng/master.py", line 79, in startService
    self._setup_amp(d)
  File "/home/km/pyng/lib/pyng/master.py", line 363, in _setup_amp
    pm.addProcess(f'amp-{i}', [runner,
  File "/usr/local/lib/python3.12/site-packages/twisted/runner/procmon.py", line 254, in addProcess
    self.startProcess(name)
  File "/usr/local/lib/python3.12/site-packages/twisted/runner/procmon.py", line 347, in startProcess
    self._reactor.spawnProcess(
  File "/usr/local/lib/python3.12/site-packages/twisted/internet/posixbase.py", line 194, in spawnProcess
    return process.Process(
  File "/usr/local/lib/python3.12/site-packages/twisted/internet/process.py", line 820, in __init__
    self._fork(path, uid, gid, executable, args, environment, fdmap=fdmap)
  File "/usr/local/lib/python3.12/site-packages/twisted/internet/process.py", line 401, in _fork
    if self._trySpawnInsteadOfFork(
  File "/usr/local/lib/python3.12/site-packages/twisted/internet/process.py", line 890, in _trySpawnInsteadOfFork
    self.pid = os.posix_spawnp(
builtins.BlockingIOError: [Errno 35] Resource temporarily unavailable: '/home/km/pyng/bin/amp-runner.py'
```

Commit:
ProccessMinotor: make startProcess() catch exceptions raised by spawnProcess().

juanpeha committed:
Merge branch 'trunk' into issue12421

Commit:
Make flake8 happy.

Commit:
Test coverage.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adiroiban commented on src/twisted/runner/procmon.py:
```diff
@@ -344,15 +344,18 @@ def startProcess(self, name):
+                env=process.env,
+                path=process.cwd,
+            )
+        except OSError:
```
Is is ok to discard the reason / details of this error?

adiroiban replied:
Also, do we want to retry on any OSError or only just on "temporary" errors like `Resource temporarily unavailable` ?

Should we retry if we get a permission denied error or a too many open files errors ?

glyph replied:
Also, is this even being handled in the right place? `IReactorProcess.spawnProcess` does not document a `@raises OSError`, which suggests that perhaps we ought to be delivering this error to `processExited` rather than raising it synchronously?

It's been raising `OSError` for a long time of course, maybe this is just a documentation issue that needs to be addressed, but I bet that this is, like, `WindowsError` sometimes instead

glyph replied:
I think that this was addressed by including it in the failure message (although as described there, I suggested `Logger.failure`)

adiroiban commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -102,6 +102,9 @@ def __init__(self):
         self.spawnedProcesses = []
 
+        # set flag to make spawnProcess OSError
+        self._spawn_resource_unavailable = False
```
can we have this a public member and document it in a similar way to `@ivar spawnedProcesses` ?

I guess that a boolean will work for now.
In the future, I guess that we might want to pass a list of Exceptions, each exception being removed from the list and then raised

glyph replied:
just a coding-standard compliance suggestion:

```suggestion
        self._spawnResourceUnavailable = False
```

adiroiban commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -279,6 +284,16 @@ def test_startProcessUnknownKeyError(self):
+        self.reactor._spawn_resource_unavailable = True
+        self.pm.addProcess("foo", ["foo"])
+        self.pm.startProcess("foo")
+        self.assertEquals(self.pm.delay["foo"], 4)
```
This test is hard to read.

Maybe update the docstring of the test to explain what is going on here, or add an comment
```suggestion
        # The process is not started,
        # instead a restart is scheduled in 4 seconds,
        # where 4 seconds is the initial minumum interval.
        self.assertEquals(self.pm.delay["foo"], 4)
```

btw... it would be nice to have the value `4` part of the setup of this test, to make it easier to understand what that value 4 stand for.

adiroiban reviewed (commented):
Thanks for the changes.

They look good.

I left a few small comments.

My main comment is about retrying with droping the error details.

Maybe we should emit a log.

Thanks again for the PR

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -279,6 +284,16 @@ def test_startProcessUnknownKeyError(self):
+        L{IReactorProcess.spawnProcess} might raise C{OSError}, ensure it
+        is caught and process is restarted
+        """
+        self.reactor._spawn_resource_unavailable = True
```
As @adiroiban pointed out, this should really be a public API on the control surface of the reactor.  Given that the class is just within this one test file, it should be fine to just add it as a method.

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -118,6 +121,8 @@ def spawnProcess(
         arguments and returns a L{DummyProcess}.
         """
+        if self._spawn_resource_unavailable:
+            raise BlockingIOError("[Errno 35] Resource temporarily unavailable")
```
You're using the constructor for `BlockingIOError` incorrectly here.  What you want is `BlockingIOError(errno.EAGAIN, os.strerror(EAGAIN))` .  You should use the errno so nobody needs "35" memorized.

But also… `EAGAIN` *in particular* seems like a weird error for `spawnProcess` not to handle itself. It seems like a bug; in other async operations it's Twisted's job to handle EWOULDBLOCK / EAGAIN (etc) for you so you don't need to think about it. When and why does this happen?

glyph reviewed (commented):
Thanks so much for contributing this fix! It seems to me that there might need to be a bit more consideration of whether this is the right place to fix this bug, though, since it's not entirely clear to me what the significance of this particular error is.

Commit:
Add `reason` parameter to connectionLost, and pass exception from startProcess to connectionLost to handle there. Adjust tests.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph commented:
> As for `OSError` being raised by `spawnProcess`, I read it in https://docs.twisted.org/en/stable/api/twisted.internet.interfaces.IReactorProcess.html#spawnProcess.

Oh, there it is, plain as day. I just didn't see it. Great!

juanpeha commented on src/twisted/runner/procmon.py:
```diff
@@ -286,7 +287,7 @@ def stopService(self):
             self.stopProcess(name)
 
-    def connectionLost(self, name):
+    def connectionLost(self, name, reason):
```
Please suggest what would be a suitable way to handle `reason` in connectionLost.

juanpeha replied:
@glyph @adiroiban As I am not familiar with Twisted internals (or logging, yet, for that matter), please suggest what would be appropriate way to handle `reason`.

adiroiban replied:
Maybe we should deprecate the `connectionLost` method.
Since it's not following the same signature as the other `connectionLost` implementation, it can be confusing.

We can create a new `processExit()` method that is explicit and might also receive the exit code.

Commit:
Failure() will use use the current exception by default.

adiroiban commented:
> As I am not familiar with Twisted internals (or logging, yet, for that matter), please suggest what would be appropriate way to handle reason.

Thanks for the updates.

I have not used the ProccessMonitor code.

I don't think that we have to many options here.

The "standard" [connectionLost](https://docs.twistedmatrix.com/en/stable/api/twisted.internet.protocol.Protocol.html#connectionLost) needs a Failure as the argument

Here, I see that we pass a "name"


-----

So I guess that the only option is to log the error details.

The documentation for Twisted logger is here

https://docs.twistedmatrix.com/en/stable/core/howto/logger.html

I think that this can be logges as a `warn` level.

-------

Just a quick feedback.

As I have not used this code, I can't provide a better feedback or guidance as I have no idea what are the use cases for this code. Sorry

Commit:
Better exit handling.

Revert connectionLost to having the name parameter, and make it call newly
added processExit, which has name and reason as parameters.

Log if process is exiting for some other reason than ProcessDone.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
s/connectionLost/processExit/ in test docstrings.

Commit:
Ignore connectionLost -> processExit call from coverage reporting.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph commented on src/twisted/runner/procmon.py:
```diff
@@ -302,6 +304,31 @@ def connectionLost(self, name):
         """
+        return self.processExit(name)  # pragma: no cover
+
+    def processExit(self, name, reason=None):
```
Does this method need to be public? It seems like this is for internal bookkeeping, not for users to call; in general we should not be adding additional API surface for users to understand if they're not supposed to interact with it.

glyph commented on src/twisted/runner/procmon.py:
```diff
@@ -302,6 +304,31 @@ def connectionLost(self, name):
+        @param name: A string that uniquely identifies the process
+            which exited.
+        @type reason: C{Failure}
+        @param reason: The reason why the connection was lost.
```
These `@type` documentation fields should be type annotations instead.

glyph commented on src/twisted/runner/procmon.py:
```diff
@@ -302,6 +304,31 @@ def connectionLost(self, name):
+        if reason and not reason.check(error.ProcessDone):
+            self.log.warn(
+                "Process '{name}' has exited: {reason!r}", name=name, reason=reason
+            )
```
Given that we do not know exactly why this process has exited (`OSError` can mean a lot of different things) I think `log.failure` might be more appropriate.  I think we should probably work *towards* something at a `log.warn` level, something that doesn't need a traceback, via (for example) enumerating specific `errno` values that we expect and mean something specific, but as long as we're hanging on to a full `Failure` here we should use the standard channel for such things.

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -94,6 +96,9 @@ class DummyProcessReactor(MemoryReactor, Clock):
     @type spawnedProcesses: C{list}
+
+    @ivar spawnProcessException: An exception which spawnProcess() will raise.
+    @type spawnProcessException: C{Exception}
```
This `@type` should be a type annotation; also, it's not correct, because it's `Exception | None`.

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -102,6 +107,8 @@ def __init__(self):
 
         self.spawnedProcesses = []
 
+        self.spawnProcessException = None
```
```suggestion
        self.spawnProcessException: Exception | None = None
```

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -279,6 +288,30 @@ def test_startProcessUnknownKeyError(self):
         """
         self.assertRaises(KeyError, self.pm.startProcess, "foo")
 
+    def test_startProcessSpawnCaughtException(self):
```
New tests should have type checking.
```suggestion
    def test_startProcessSpawnCaughtException(self) -> None:
```

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -279,6 +288,30 @@ def test_startProcessUnknownKeyError(self):
+        # process will be restarted in 123 seconds, per minRestartDelay and maxRestartDelay above
+        self.assertEquals(self.pm.delay["foo"], 123)
+
+    def test_startProcessSpawnUncaughtException(self):
```
```suggestion
    def test_startProcessSpawnUncaughtException(self) -> None:
```

glyph commented on src/twisted/runner/test/test_procmon.py:
```diff
@@ -118,6 +125,8 @@ def spawnProcess(
         Fake L{reactor.spawnProcess}, that logs all the process
         arguments and returns a L{DummyProcess}.
         """
+        if self.spawnProcessException:
```
This is mostly an academic concern, but there's no reason to invoke `__nonzero__` for a `None` check; we should be explicit (and I think the type checker would prefer it):

```suggestion
        if self.spawnProcessException is not None:
```

glyph reviewed (commented):
Almost there, I think. Just some polish around type annotations and public methods.

glyph commented:
> Would you just complete the remaining, @glyph, please?

Yep, will do.

glyph committed:
Update src/twisted/runner/test/test_procmon.py

glyph committed:
Update src/twisted/runner/test/test_procmon.py

glyph committed:
Update src/twisted/runner/test/test_procmon.py

glyph committed:
Update src/twisted/runner/test/test_procmon.py

juanpeha committed:
Merge branch 'trunk' into issue12421

glyph committed:
Merge branch 'trunk' into issue12421

glyph committed:
furnish types

glyph committed:
reformat, annotations help, make method private per review

glyph committed:
complete type coverage and removal of exception while I'm here

glyph committed:
always good to add one of these

glyph committed:
idiomatic error handling

glyph committed:
don't rely on inference (make them pydoctor legible)

glyph reviewed (approved):
I made changes that should address the rest of my feedback, so I think we can merge.
